### ISSUE TITLE
Set `.train()` or `.eval()` via `optimization.context` attribute

### ISSUE THREAD

**Issue by gideonite** [ai2cm/ace#381]
Introduce an `optimization.context` attribute onto `NullOptimization` and `Optimization` which sets `.train()` or `.eval()`

> It does have the downside that you need to remember to call `.train()` or `.eval()` every time before the model is called. I want to suggest in a different PR introducing a `optimization.context` attribute onto NullOptimization and Optimization which sets `.train()` or `.eval()` (like discussed here https://discuss.pytorch.org/t/opinion-eval-should-be-a-context-manager/18998), where Optimization sets train and NullOptimization sets eval, and then performing module operations within this context in the Model. This would make it so any inference code or notebooks loading and using your Model will also appropriately set the module's mode.

> This would fix an issue in this PR where `fme.core.downscaling.evaluator` will still run with dropout, and prevent any similar issues in any other similar scripts.

_Originally posted by @mcgibbon in https://github.com/ai2cm/full-model/pull/972#pullrequestreview-2163597087_

**Comment by mcgibbon** [comment]
I edited down the issue title to a shorter length for the issues list and moved the original title into the description.

**Comment by gideonite** [comment]
I've started working on this in ai2cm/full-model#989 but I'm not sure if this is what you are proposing. I suppose using a context is more likely to be flagged by future developers as required when using the optimization class but I'm also not sure if we want to add more stringent constraints to somehow ensure that mode is set properly for training and evaluation.

https://github.com/ai2cm/full-model/blob/00610f2b4409e53740fd59200589042b51bea273/fme/fme/core/optimization.py#L43-L50

**Comment by mcgibbon** [comment]
Close, except Optimization will always set train and NullOptimization will always set eval. See the existing set_mode routine.

**Comment by mcgibbon** [comment]
Alternatively, you could call the set mode routine (in the run/train/generate methods) without making it a context manager.

I had forgotten this method exists.
